File: backend/pkg/gameplugin/sdk/game-plugin/src/network.ts
import { Client, MessageOption } from './client';
import { HOST_API_STATUS_SUCCESS, invokeHostAPI } from './hostapi';

export const HOST_NETWORK_METHOD = {
  REQUEST: 'host.network.request',
  TCP_OPEN: 'host.network.tcp.open',
  TCP_READ: 'host.network.tcp.read',
  TCP_WRITE: 'host.network.tcp.write',
  TCP_CLOSE: 'host.network.tcp.close',
  UDP_OPEN: 'host.network.udp.open',
  UDP_RECEIVE: 'host.network.udp.receive',
  UDP_SEND: 'host.network.udp.send',
  UDP_CLOSE: 'host.network.udp.close',
  WEBSOCKET_OPEN: 'host.network.websocket.open',
  WEBSOCKET_RECEIVE: 'host.network.websocket.receive',
  WEBSOCKET_SEND: 'host.network.websocket.send',
  WEBSOCKET_CLOSE: 'host.network.websocket.close',
} as const;

export interface NetworkRequestInput {
  method?: string;
  url: string;
  headers?: Record<string, string>;
  bodyBase64?: string;
  timeoutMs?: number;
  maxResponseBytes?: number;
}

export interface NetworkRequestOutput {
  statusCode: number;
  headers: Record<string, string[]>;
  bodyBase64: string;
  finalUrl: string;
}

export interface NetworkSocketOpenInput {
  target: string;
  port: number;
  timeoutMs?: number;
}

export interface NetworkSocketOpenOutput {
  handleId: string;
  transport: string;
  localAddress?: string;
  remoteAddress?: string;
}

export interface NetworkSocketReadInput {
  handleId: string;
  maxBytes?: number;
  timeoutMs?: number;
}

export interface NetworkSocketReadOutput {
  dataBase64: string;
  bytesRead: number;
  eof?: boolean;
}

export interface NetworkSocketWriteInput {
  handleId: string;
  dataBase64: string;
  timeoutMs?: number;
}

export interface NetworkSocketWriteOutput {
  bytesWritten: number;
}

export interface NetworkSocketCloseInput {
  handleId: string;
}

export interface NetworkSocketCloseOutput {
  closed: boolean;
}

export interface NetworkWebSocketOpenInput {
  url: string;
  headers?: Record<string, string>;
  subprotocols?: string[];
  timeoutMs?: number;
}

export interface NetworkWebSocketOpenOutput {
  handleId: string;
  subprotocol?: string;
  remoteAddress?: string;
}

export interface NetworkWebSocketSendInput {
  handleId: string;
  messageType?: 'text' | 'binary';
  dataBase64: string;
  timeoutMs?: number;
}

export interface NetworkWebSocketSendOutput {
  bytesWritten: number;
}

export interface NetworkWebSocketReceiveInput {
  handleId: string;
  timeoutMs?: number;
}

export interface NetworkWebSocketReceiveOutput {
  messageType: 'text' | 'binary';
  dataBase64: string;
  bytesRead: number;
}

async function invokeNetwork<T>(
  client: Client,
  method: string,
  input: unknown,
  opts: MessageOption[] = [],
): Promise<T> {
  const result = await invokeHostAPI(client, { method, version: 1, input }, opts);
  if (result.status !== HOST_API_STATUS_SUCCESS) {
    throw new Error(`Game Plugin SDK: ${method} returned host API status '${result.status}'`);
  }
  return result.output as T;
}

export function networkRequest(client: Client, input: NetworkRequestInput, opts: MessageOption[] = []): Promise<NetworkRequestOutput> {
  return invokeNetwork(client, HOST_NETWORK_METHOD.REQUEST, input, opts);
}

export function networkTCPOpen(client: Client, input: NetworkSocketOpenInput, opts: MessageOption[] = []): Promise<NetworkSocketOpenOutput> {
  return invokeNetwork(client, HOST_NETWORK_METHOD.TCP_OPEN, input, opts);
}

export function networkTCPRead(client: Client, input: NetworkSocketReadInput, opts: MessageOption[] = []): Promise<NetworkSocketReadOutput> {
  return invokeNetwork(client, HOST_NETWORK_METHOD.TCP_READ, input, opts);
}

export function networkTCPWrite(client: Client, input: NetworkSocketWriteInput, opts: MessageOption[] = []): Promise<NetworkSocketWriteOutput> {
  return invokeNetwork(client, HOST_NETWORK_METHOD.TCP_WRITE, input, opts);
}

export function networkTCPClose(client: Client, input: NetworkSocketCloseInput, opts: MessageOption[] = []): Promise<NetworkSocketCloseOutput> {
  return invokeNetwork(client, HOST_NETWORK_METHOD.TCP_CLOSE, input, opts);
}

export function networkUDPOpen(client: Client, input: NetworkSocketOpenInput, opts: MessageOption[] = []): Promise<NetworkSocketOpenOutput> {
  return invokeNetwork(client, HOST_NETWORK_METHOD.UDP_OPEN, input, opts);
}

export function networkUDPReceive(client: Client, input: NetworkSocketReadInput, opts: MessageOption[] = []): Promise<NetworkSocketReadOutput> {
  return invokeNetwork(client, HOST_NETWORK_METHOD.UDP_RECEIVE, input, opts);
}

export function networkUDPSend(client: Client, input: NetworkSocketWriteInput, opts: MessageOption[] = []): Promise<NetworkSocketWriteOutput> {
  return invokeNetwork(client, HOST_NETWORK_METHOD.UDP_SEND, input, opts);
}

export function networkUDPClose(client: Client, input: NetworkSocketCloseInput, opts: MessageOption[] = []): Promise<NetworkSocketCloseOutput> {
  return invokeNetwork(client, HOST_NETWORK_METHOD.UDP_CLOSE, input, opts);
}

export function networkWebSocketOpen(client: Client, input: NetworkWebSocketOpenInput, opts: MessageOption[] = []): Promise<NetworkWebSocketOpenOutput> {
  return invokeNetwork(client, HOST_NETWORK_METHOD.WEBSOCKET_OPEN, input, opts);
}

export function networkWebSocketReceive(client: Client, input: NetworkWebSocketReceiveInput, opts: MessageOption[] = []): Promise<NetworkWebSocketReceiveOutput> {
  return invokeNetwork(client, HOST_NETWORK_METHOD.WEBSOCKET_RECEIVE, input, opts);
}

export function networkWebSocketSend(client: Client, input: NetworkWebSocketSendInput, opts: MessageOption[] = []): Promise<NetworkWebSocketSendOutput> {
  return invokeNetwork(client, HOST_NETWORK_METHOD.WEBSOCKET_SEND, input, opts);
}

export function networkWebSocketClose(client: Client, input: NetworkSocketCloseInput, opts: MessageOption[] = []): Promise<NetworkSocketCloseOutput> {
  return invokeNetwork(client, HOST_NETWORK_METHOD.WEBSOCKET_CLOSE, input, opts);
}

File: backend/pkg/gameplugin/sdk/go/network.go
package sdk

import (
	"context"
	"encoding/json"
	"fmt"
)

const (
	MethodHostNetworkRequest          = "host.network.request"
	MethodHostNetworkTCPOpen          = "host.network.tcp.open"
	MethodHostNetworkTCPRead          = "host.network.tcp.read"
	MethodHostNetworkTCPWrite         = "host.network.tcp.write"
	MethodHostNetworkTCPClose         = "host.network.tcp.close"
	MethodHostNetworkUDPOpen          = "host.network.udp.open"
	MethodHostNetworkUDPReceive       = "host.network.udp.receive"
	MethodHostNetworkUDPSend          = "host.network.udp.send"
	MethodHostNetworkUDPClose         = "host.network.udp.close"
	MethodHostNetworkWebSocketOpen    = "host.network.websocket.open"
	MethodHostNetworkWebSocketReceive = "host.network.websocket.receive"
	MethodHostNetworkWebSocketSend    = "host.network.websocket.send"
	MethodHostNetworkWebSocketClose   = "host.network.websocket.close"
)

type NetworkRequestInput struct {
	Method           string            `json:"method,omitempty"`
	URL              string            `json:"url"`
	Headers          map[string]string `json:"headers,omitempty"`
	BodyBase64       string            `json:"bodyBase64,omitempty"`
	TimeoutMs        int               `json:"timeoutMs,omitempty"`
	MaxResponseBytes int64             `json:"maxResponseBytes,omitempty"`
}

type NetworkRequestOutput struct {
	StatusCode int                 `json:"statusCode"`
	Headers    map[string][]string `json:"headers"`
	BodyBase64 string              `json:"bodyBase64"`
	FinalURL   string              `json:"finalUrl"`
}

type NetworkSocketOpenInput struct {
	Target    string `json:"target"`
	Port      int    `json:"port"`
	TimeoutMs int    `json:"timeoutMs,omitempty"`
}

type NetworkSocketOpenOutput struct {
	HandleID      string `json:"handleId"`
	Transport     string `json:"transport"`
	LocalAddress  string `json:"localAddress,omitempty"`
	RemoteAddress string `json:"remoteAddress,omitempty"`
}

type NetworkSocketReadInput struct {
	HandleID  string `json:"handleId"`
	MaxBytes  int    `json:"maxBytes,omitempty"`
	TimeoutMs int    `json:"timeoutMs,omitempty"`
}

type NetworkSocketReadOutput struct {
	DataBase64 string `json:"dataBase64"`
	BytesRead  int    `json:"bytesRead"`
	EOF        bool   `json:"eof,omitempty"`
}

type NetworkSocketWriteInput struct {
	HandleID   string `json:"handleId"`
	DataBase64 string `json:"dataBase64"`
	TimeoutMs  int    `json:"timeoutMs,omitempty"`
}

type NetworkSocketWriteOutput struct {
	BytesWritten int `json:"bytesWritten"`
}

type NetworkSocketCloseInput struct {
	HandleID string `json:"handleId"`
}

type NetworkSocketCloseOutput struct {
	Closed bool `json:"closed"`
}

type NetworkWebSocketOpenInput struct {
	URL          string            `json:"url"`
	Headers      map[string]string `json:"headers,omitempty"`
	Subprotocols []string          `json:"subprotocols,omitempty"`
	TimeoutMs    int               `json:"timeoutMs,omitempty"`
}

type NetworkWebSocketOpenOutput struct {
	HandleID      string `json:"handleId"`
	Subprotocol   string `json:"subprotocol,omitempty"`
	RemoteAddress string `json:"remoteAddress,omitempty"`
}

type NetworkWebSocketSendInput struct {
	HandleID    string `json:"handleId"`
	MessageType string `json:"messageType,omitempty"`
	DataBase64  string `json:"dataBase64"`
	TimeoutMs   int    `json:"timeoutMs,omitempty"`
}

type NetworkWebSocketSendOutput struct {
	BytesWritten int `json:"bytesWritten"`
}

type NetworkWebSocketReceiveInput struct {
	HandleID  string `json:"handleId"`
	TimeoutMs int    `json:"timeoutMs,omitempty"`
}

type NetworkWebSocketReceiveOutput struct {
	MessageType string `json:"messageType"`
	DataBase64  string `json:"dataBase64"`
	BytesRead   int    `json:"bytesRead"`
}

func invokeTypedHostNetworkAPI(ctx context.Context, client *Client, method string, input any, output any, opts ...MessageOption) error {
	if client == nil {
		return fmt.Errorf("game plugin sdk: client is required")
	}
	encoded, err := json.Marshal(input)
	if err != nil {
		return NewEncodeError("marshal %s input: %v", method, err)
	}
	result, err := client.InvokeHostAPI(ctx, HostInvokeInput{Method: method, Version: 1, Input: encoded}, opts...)
	if err != nil {
		return err
	}
	if result.Status != HostAPISuccess {
		return fmt.Errorf("game plugin sdk: %s returned host API status %q", method, result.Status)
	}
	if output == nil || len(result.Output) == 0 {
		return nil
	}
	if err := json.Unmarshal(result.Output, output); err != nil {
		return NewEncodeError("unmarshal %s output: %v", method, err)
	}
	return nil
}

func NetworkRequest(ctx context.Context, client *Client, input NetworkRequestInput, opts ...MessageOption) (NetworkRequestOutput, error) {
	var out NetworkRequestOutput
	err := invokeTypedHostNetworkAPI(ctx, client, MethodHostNetworkRequest, input, &out, opts...)
	return out, err
}

func NetworkTCPOpen(ctx context.Context, client *Client, input NetworkSocketOpenInput, opts ...MessageOption) (NetworkSocketOpenOutput, error) {
	var out NetworkSocketOpenOutput
	err := invokeTypedHostNetworkAPI(ctx, client, MethodHostNetworkTCPOpen, input, &out, opts...)
	return out, err
}

func NetworkTCPRead(ctx context.Context, client *Client, input NetworkSocketReadInput, opts ...MessageOption) (NetworkSocketReadOutput, error) {
	var out NetworkSocketReadOutput
	err := invokeTypedHostNetworkAPI(ctx, client, MethodHostNetworkTCPRead, input, &out, opts...)
	return out, err
}

func NetworkTCPWrite(ctx context.Context, client *Client, input NetworkSocketWriteInput, opts ...MessageOption) (NetworkSocketWriteOutput, error) {
	var out NetworkSocketWriteOutput
	err := invokeTypedHostNetworkAPI(ctx, client, MethodHostNetworkTCPWrite, input, &out, opts...)
	return out, err
}

func NetworkTCPClose(ctx context.Context, client *Client, input NetworkSocketCloseInput, opts ...MessageOption) (NetworkSocketCloseOutput, error) {
	var out NetworkSocketCloseOutput
	err := invokeTypedHostNetworkAPI(ctx, client, MethodHostNetworkTCPClose, input, &out, opts...)
	return out, err
}

func NetworkUDPOpen(ctx context.Context, client *Client, input NetworkSocketOpenInput, opts ...MessageOption) (NetworkSocketOpenOutput, error) {
	var out NetworkSocketOpenOutput
	err := invokeTypedHostNetworkAPI(ctx, client, MethodHostNetworkUDPOpen, input, &out, opts...)
	return out, err
}

func NetworkUDPReceive(ctx context.Context, client *Client, input NetworkSocketReadInput, opts ...MessageOption) (NetworkSocketReadOutput, error) {
	var out NetworkSocketReadOutput
	err := invokeTypedHostNetworkAPI(ctx, client, MethodHostNetworkUDPReceive, input, &out, opts...)
	return out, err
}

func NetworkUDPSend(ctx context.Context, client *Client, input NetworkSocketWriteInput, opts ...MessageOption) (NetworkSocketWriteOutput, error) {
	var out NetworkSocketWriteOutput
	err := invokeTypedHostNetworkAPI(ctx, client, MethodHostNetworkUDPSend, input, &out, opts...)
	return out, err
}

func NetworkUDPClose(ctx context.Context, client *Client, input NetworkSocketCloseInput, opts ...MessageOption) (NetworkSocketCloseOutput, error) {
	var out NetworkSocketCloseOutput
	err := invokeTypedHostNetworkAPI(ctx, client, MethodHostNetworkUDPClose, input, &out, opts...)
	return out, err
}

func NetworkWebSocketOpen(ctx context.Context, client *Client, input NetworkWebSocketOpenInput, opts ...MessageOption) (NetworkWebSocketOpenOutput, error) {
	var out NetworkWebSocketOpenOutput
	err := invokeTypedHostNetworkAPI(ctx, client, MethodHostNetworkWebSocketOpen, input, &out, opts...)
	return out, err
}

func NetworkWebSocketReceive(ctx context.Context, client *Client, input NetworkWebSocketReceiveInput, opts ...MessageOption) (NetworkWebSocketReceiveOutput, error) {
	var out NetworkWebSocketReceiveOutput
	err := invokeTypedHostNetworkAPI(ctx, client, MethodHostNetworkWebSocketReceive, input, &out, opts...)
	return out, err
}

func NetworkWebSocketSend(ctx context.Context, client *Client, input NetworkWebSocketSendInput, opts ...MessageOption) (NetworkWebSocketSendOutput, error) {
	var out NetworkWebSocketSendOutput
	err := invokeTypedHostNetworkAPI(ctx, client, MethodHostNetworkWebSocketSend, input, &out, opts...)
	return out, err
}

func NetworkWebSocketClose(ctx context.Context, client *Client, input NetworkSocketCloseInput, opts ...MessageOption) (NetworkSocketCloseOutput, error) {
	var out NetworkSocketCloseOutput
	err := invokeTypedHostNetworkAPI(ctx, client, MethodHostNetworkWebSocketClose, input, &out, opts...)
	return out, err
}
